checksum/runner_service.py

```diff
@@ -0,0 +1,268 @@
+    queue is full, the oldest job is removed (provided it is not still
+    running).
+
+    Methods
+    -------
+    start(cmd, **kwargs):
+        start a new job
+    stop(job_id):
+        stop job with given id
+    stop_all:
+        stop all running jobs
+    status:
+        return the status of the job with the given id
+    status_all:
+        return status of all jobs in the history
+    """
+
+    def __init__(self, history_len=100):
+        """
+        Parameters
+        ----------
+        history_len: int
+            maximum number of jobs to keep track of.
+        """
+        self._job_history = collections.deque(maxlen=history_len)
+        self._next_id = 1
+        self._lock = asyncio.Lock()
+
+    async def _generate_next_id(self):
+        """
+        Returns a valid job id
+        """
+        async with self._lock:
+            next_id = self._next_id
+            self._next_id += 1
+            return next_id
+
+    def _get_job(self, job_id):
+        """
+        Returns the job with the given job id
+
+        Parameters
+        ----------
+        job_id: int
+            id of the desired job
+
+        Raises
+        ------
+        IndexError
+            when no job has the given id. This can mean the job never existed
+            or is too old and was removed from memory.
+
+        Returns
+        -------
+        Job
+            job with the given job id
+        """
+        try:
+            return next(
+                job
+                for job in self._job_history
+                if job.job_id == job_id)
+        except StopIteration:
+            msg = f"job {job_id} not found"
+            log.warning(msg)
+            raise IndexError(msg)
+
+    async def start(self, cmd, **kwargs):
+        """
+        Start executing a new command.
+
+        Parameters
+        ----------
+        cmd: [str]
+            command to be executed
+        **kwargs:
+            keyword arguments to be forwarded to subprocess.Popen
+
+        Raises
+        ------
+        RuntimeError
+            if the history is full and the oldest job is still running
+
+        Returns
+        -------
+        job_id: int
+        """
+        if (
+            self._job_history.maxlen == len(self._job_history)
+            and self._job_history[-1].get_status() == arteria_state.STARTED
+        ):
+            msg = (
+                "Could not start a new job because the history is full "
+                "and the oldest job is still running."
+                )
+            log.error(msg)
+            raise RuntimeError(msg)
+
+        job_id = await self._generate_next_id()
+        job = Job(job_id, cmd, **kwargs)
+
+        async with self._lock:
+            self._job_history.appendleft(job)
+
+        return job.job_id
+
+    def stop(self, job_id):
+        """
+        Stop the job with the given id.
+
+        If no job with the given id is found, nothing is done.
```

Comment:
No really necessary, but it would be user friendly to report back if the id can't be found.

Reply:
I now realize it might be a good idea to keep it like this to not change the interface, like you mention in the PR description.